cs/VowpalWabbitAsync.cs

```diff
@@ -96,16 +170,54 @@ private void Dispose(bool disposing)
         }
     }
 
+    /// <summary>
+    /// An async VW wrapper for multiline ingest.
+    /// </summary>
+    /// <typeparam name="TExample">The user type of the shared feature.</typeparam>
+    /// <typeparam name="TActionDependentFeature">The user type for each action dependent feature.</typeparam>
     public class VowpalWabbitAsync<TExample, TActionDependentFeature> : IDisposable
     {
+        /// <summary>
+        /// The owning manager.
+        /// </summary>
         private readonly VowpalWabbitThreadedLearning manager;
 
+        /// <summary>
+        /// The serializers are not thread-safe. Thus we need to allocate one for each VW instance.
+        /// </summary>
         private VowpalWabbitSerializer<TExample>[] serializers;
 
+        /// <summary>
+        /// The serializers are not thread-safe. Thus we need to allocate one for each VW instance.
+        /// </summary>
         private VowpalWabbitSerializer<TActionDependentFeature>[] actionDependentFeatureSerializers;
 
         internal VowpalWabbitAsync(VowpalWabbitThreadedLearning manager)
         {
+            if (manager == null)
+            {
+                throw new ArgumentNullException("manager");
```

Comment:
Contract.Requires doesn't throw always (just if compiled correctly). I use the direct checks in places that aren't on the hot path